arwen/contexts/metering.go

```diff
@@ -99,20 +105,187 @@ func (context *meteringContext) PopDiscard() {
 	context.stateStack = context.stateStack[:stateStackLen-1]
 }
 
+// PopMergeActiveState pops the state at the top of the internal stack and
+// merges it into the active state
+func (context *meteringContext) PopMergeActiveState() {
+	stateStackLen := len(context.stateStack)
+	if stateStackLen == 0 {
+		return
+	}
+
+	prevState := context.stateStack[stateStackLen-1]
+	context.stateStack = context.stateStack[:stateStackLen-1]
+
+	context.initialGasProvided = prevState.initialGasProvided
+	context.initialCost = prevState.initialCost
+	context.gasForExecution = prevState.gasForExecution
+
+	context.addToGasUsedByAccounts(prevState.gasUsedByAccounts)
+}
+
+func (context *meteringContext) cloneGasUsedByAccounts() map[string]uint64 {
+	clone := make(map[string]uint64, len(context.gasUsedByAccounts))
+
+	for address, gasUsed := range context.gasUsedByAccounts {
+		clone[address] = gasUsed
+	}
+
+	return clone
+}
+
+func (context *meteringContext) addToGasUsedByAccounts(gasUsed map[string]uint64) {
+	for address, gas := range gasUsed {
+		context.gasUsedByAccounts[address] += gas
+	}
+}
+
+// UpdateGasStateOnSuccess performs final gas accounting after a successful execution.
+func (context *meteringContext) UpdateGasStateOnSuccess(vmOutput *vmcommon.VMOutput) error {
+	context.updateSCGasUsed()
+	err := context.setGasUsedToOutputAccounts(vmOutput)
+	if err != nil {
+		return err
+	}
+
+	err = context.checkGas(vmOutput)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// UpdateGasStateOnSuccess performs final gas accounting after a failed execution.
+func (context *meteringContext) UpdateGasStateOnFailure(vmOutput *vmcommon.VMOutput) {
+	runtime := context.host.Runtime()
+	output := context.host.Output()
+
+	account, _ := output.GetOutputAccount(runtime.GetSCAddress())
+	account.GasUsed = math.AddUint64(account.GasUsed, context.GetGasProvided())
+}
+
+func (context *meteringContext) updateSCGasUsed() {
+	runtime := context.host.Runtime()
+	output := context.host.Output()
+
+	currentAccountAddress := runtime.GetSCAddress()
+	currentContractAccount, _ := output.GetOutputAccount(currentAccountAddress)
+	outputAccounts := context.host.Output().GetOutputAccounts()
+
+	gasTransferredByCurrentAccount := context.getGasTransferredByAccount(currentContractAccount)
+	gasUsedByOthers := context.getGasUsedByAllOtherAccounts(outputAccounts)
+
+	gasUsed := context.GasSpentByContract()
+	gasUsed = math.SubUint64(gasUsed, gasTransferredByCurrentAccount)
+	gasUsed = math.SubUint64(gasUsed, gasUsedByOthers)
+
+	context.gasUsedByAccounts[string(currentAccountAddress)] = gasUsed
+}
+
+func (context *meteringContext) TrackGasUsedByBuiltinFunction(
+	err error,
+	builtinInput *vmcommon.ContractCallInput,
+	builtinOutput *vmcommon.VMOutput,
+	postBuiltinInput *vmcommon.ContractCallInput,
+) {
+	if err != nil {
```

Comment:
I think this could be called without the error input. and copy simply the useGas function wherever is needed. Only 3 places as I saw.

Reply:
Agreed, changed accordingly.